operating_unit_grt_api_sync/models/operating_unit.py

```diff
@@ -0,0 +1,182 @@
+            "code": code,
+            "company_id": company_id,
+            "valid_from": data["valid_from"],
+            "valid_until": data["valid_until"],
+            "synced_with_grt": True,
+        }
+
+    def _prepare_branch_partner_create_vals(self, data, country_mapping):
+        branch_name = self._prepare_branch_name(data["branch"])
+        country_id = country_mapping[data["country"]]
+        return {
+            "name": f"OU {data['operating_unit']}, {branch_name}",
+            "city": data["branch"],
+            "country_id": country_id,
+        }
+
+    def _create_new_branches(self, data, ou_code_company_mapping):
+        vals = []
+        partner_vals = []
+        countries = self.env["res.country"].search([])
+        country_mapping = {country.name: country.id for country in countries}
+        for code, branch_data in data.items():
+            create_vals = self._prepare_branch_create_vals(
+                code, branch_data, ou_code_company_mapping
+            )
+            vals.append(create_vals)
+            partner_create_vals = self._prepare_branch_partner_create_vals(
+                branch_data, country_mapping
+            )
+            partner_vals.append(partner_create_vals)
+        partners = self.env["res.partner"].create(partner_vals)
+        vals_with_partner = []
+        for partner, data in zip(partners, vals):
+            data.update({"partner_id": partner.id})
+            vals_with_partner.append(data)
+        self.create(vals_with_partner)
+
+    @staticmethod
+    def _filter_branches_data(data, ou_code_company_mapping):
+        return [
+            d
+            for d in data
+            if d["management_id"][:3] in ou_code_company_mapping
+            and d["management_id_level_number"] == 5
+        ]
+
+    @staticmethod
+    def _filter_branches_to_update(recs, data):
+        """Filter the recordset to only include records that need to be updated."""
+        return recs.filtered(
+            lambda mid: mid.valid_from != data[mid.code]["operational_from"]
+            or mid.valid_until != data[mid.code]["operational_until"]
+            or not mid.synced_with_grt
+        )
+
+    def _update_branch(self, data):
+        self.ensure_one()
+        vals = {}
+        if self.valid_from != data["valid_from"]:
+            vals["valid_from"] = data["valid_from"]
+        if self.valid_until != data["valid_until"]:
+            vals["valid_until"] = data["valid_until"]
+        if not self.synced_with_grt:
+            vals["synced_with_grt"] = True
+        if vals:
+            self.write(vals)
+
+    def _process_grt_branch_data(self, data):
+        compatible_companies = self.env["res.company"].search(
+            [("grt_code_prefix", "!=", False)]
+        )
+        ou_code_company_mapping = {
+            c.grt_code_prefix: c.id for c in compatible_companies
+        }
+
+        if data := self._filter_branches_data(  # noqa: E231 E701
```

Comment:
I don't see benefits from filtering them in that way.
you could simply go in loop and do check before preparing branches:

```suggestion
        for d in data:
            if d.get("management_id_level_number", 0) == 5 and d.get("management_id", "")[:3] in ou_code_company_mapping:
                api_data = {...
```